lib/logstash/outputs/sqs.rb

```diff
@@ -1,136 +1,168 @@
 
-  public
-  def aws_service_endpoint(region)
-    return {
-        :sqs_endpoint => "sqs.#{region}.amazonaws.com"
-    }
-  end
+  # The name of the SQS queue to push messages into.
+  config :queue, :validate => :string, :required => true
 
-  public 
+  public
   def register
-    require "aws-sdk"
-
-    @sqs = AWS::SQS.new(aws_options_hash)
+    @sqs = Aws::SQS::Client.new(aws_options_hash)
 
     if @batch
       if @batch_events > 10
-        raise RuntimeError.new(
-          "AWS only allows a batch_events parameter of 10 or less"
-        )
+        raise LogStash::ConfigurationError, 'The maximum batch size is 10 events'
       elsif @batch_events <= 1
-        raise RuntimeError.new(
-          "batch_events parameter must be greater than 1 (or its not a batch)"
-        )
+        raise LogStash::ConfigurationError, 'The batch size must be greater than 1'
       end
-      buffer_initialize(
-        :max_items => @batch_events,
-        :max_interval => @batch_timeout,
-        :logger => @logger
-      )
     end
 
     begin
-      @logger.debug("Connecting to AWS SQS queue '#{@queue}'...")
-      @sqs_queue = @sqs.queues.named(@queue)
-      @logger.info("Connected to AWS SQS queue '#{@queue}' successfully.")
-    rescue Exception => e
-      @logger.error("Unable to access SQS queue '#{@queue}': #{e.to_s}")
-    end # begin/rescue
-  end # def register
+      @logger.debug('Connecting to AWS SQS queue', :queue => @queue, :region => region)
+      @queue_url = @sqs.get_queue_url(:queue_name => @queue)[:queue_url]
+      @logger.info('Connected to AWS SQS queue successfully', :queue => @queue, :region => region)
+    rescue Aws::SQS::Errors::ServiceError => e
+      @logger.error('Failed to connect to Amazon SQS', :error => e)
+      raise LogStash::ConfigurationError, 'Verify the SQS queue name and your credentials'
+    end
+  end
 
   public
-  def receive(event)
+  def multi_receive_encoded(encoded_events)
     if @batch
-      buffer_receive(event.to_json)
-      return
+      multi_receive_encoded_batch(encoded_events)
+    else
+      multi_receive_encoded_single(encoded_events)
+    end
+  end
+
+  private
+  def multi_receive_encoded_batch(encoded_events)
+    bytes = 0
+    entries = []
+
+    # Split the events into multiple batches to ensure that a single batch does
+    # not exceed `@message_max_size`.
+    #
+    # TODO: We can probably do this more effectively. The current
+    # implementation simply ensures that the batch size is less than some
+    # threshold, but does not ensure that batch size itself is maximized.
+    # Maximizing the size of each batch would minimize the number of batch
+    # sends that need to be performed.
+    encoded_events.each_with_index.map do |event_and_data, index|
+      event, data = event_and_data
+
+      if data.bytesize > @message_max_size
+        @logger.warn('Message exceeds maximum length and will be dropped', :message_size => data.bytesize)
+        next
+      end
```

Comment:
Logging the dropped message does seem like a good idea, but can also cause more problems than it solves. In our setup, we process Logstash's own logs back through Logstash, so outputting the entire message to Logstash's logs will cause a subsequent message to be dropped, and so on.